### PR TITLE
Enhancements & Fixes - v1

### PR DESCRIPTION
**Pull request by viraj124**

### CODE REVIEW

File: contracts/RentNft.sol [hunk 1]
```diff
@@ -130,8 +137,9 @@ contract RentNft is ReentrancyGuard, Ownable {
   ) public nonReentrant {
     Nft storage nft = nfts[_nftAddress][_tokenId];
 
+    require(nft.lender != address(0), "could not find an NFT");
     require(_borrower != nft.lender, "can't borrow own nft");
-    require(_borrower > address(0), "could not find an NFT");
+    require(_actualDuration <= nft.maxDuration, "Max Duration exceeded");
```
Comment (nazariyv) [inline contracts/RentNft.sol:142]: just being overly pedantic, I'd like to make it lower case like the other strings. I know it sounds stupid, but I think all our messages are lower case

Comment (viraj124) [reply contracts/RentNft.sol:142]: sure any other changes ?

File: contracts/RentNft.sol [hunk 2]
```diff
@@ -180,6 +190,8 @@ contract RentNft is ReentrancyGuard, Ownable {
     Nft storage nft = nfts[_nftAddress][_tokenId];
 
     require(nft.borrower == msg.sender, "not borrower");
+    uint256 durationInDays = now.sub(nft.borrowedAt).div(86400);
+    require(durationInDays <= nft.actualDuration, "duration exceeded");
```
Comment (nazariyv) [inline contracts/RentNft.sol:194]: nice

File: contracts/RentNftAddressProvider.sol [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+  function setUsdc(address _usdc) public onlyOwner {
+    _setAddress(keccak256(abi.encodePacked("USDC", networkId)), _usdc);
+    emit UsdcUpdated(_usdc);
+  }
+
+  function setUsdt(address _usdt) public onlyOwner {
+    _setAddress(keccak256(abi.encodePacked("USDT", networkId)), _usdt);
+    emit UsdtUpdated(_usdt);
+  }
```
Comment (nazariyv) [inline contracts/RentNftAddressProvider.sol:43]: @CodinMaster @viraj124 should we add any other coins here?

File: contracts/mocks/PaymentToken.sol [hunk 4]
```diff
@@ -5,6 +5,6 @@ import "@openzeppelin/contracts/token/ERC20/ERC20.sol";
 
 contract PaymentToken is ERC20 {
   constructor() public ERC20("DAI", "DAI") {
-    _mint(address(msg.sender), 100000000000000000000000);
+    _mint(address(msg.sender), 10000000000000000000000000);
```
Comment (nazariyv) [inline contracts/mocks/PaymentToken.sol:8]: couple more zeros?

Comment (viraj124) [reply contracts/mocks/PaymentToken.sol:8]: @CodinMaster added it initially while cc'ing him

File: contracts/RentNftAddressProvider.sol (hunk 3 above)
Comment (viraj124) [reply contracts/RentNftAddressProvider.sol:43]: should be fine for now I think

File: contracts/mocks/PaymentToken.sol (hunk 4 above)
Comment (viraj124) [reply contracts/mocks/PaymentToken.sol:8]: should not make a diff in testing

File: contracts/mocks/TestERC1155.sol [hunk 5]
```diff
@@ -0,0 +1,10 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.6.0;
+
+import "../../node_modules/@openzeppelin/contracts/token/ERC1155/ERC1155.sol";
```
Comment (nazariyv) [inline contracts/mocks/TestERC1155.sol:4]: this file is not supposed to be here

File: mapping.ts [hunk 6]
```diff
@@ -13,149 +13,247 @@ import {
...
   nft.borrowedAt = BigInt.fromI32(0);
+  return nft;
+};
+
+const getFaceId = (nftAddr: string, tokenId: string): string =>
+  `${nftAddr}::${tokenId}`;
+const getNftId = (faceId: string, lender: string): string =>
+  `${faceId}::${lender}`;
+const getApprovedOneId = (
+  nftAddress: string,
```
Comment (nazariyv) [inline mapping.ts:38]: this is an old mapping.ts. This will not work

File: yarn.lock [hunk 7]
```diff
@@ -226,11 +226,55 @@
   dependencies:
     assemblyscript "https://github.com/AssemblyScript/assemblyscript#36040d5b5312f19a025782b5e36663823494c2f3"
 
+"@openzeppelin/contract-loader@^0.6.1", "@openzeppelin/contract-loader@^0.6.2":
```
Comment (nazariyv) [inline yarn.lock:229]: is this correct? we have way too many deps in here

File: contracts/mocks/TestERC1155.sol (hunk 5 above)
Comment (viraj124) [reply contracts/mocks/TestERC1155.sol:4]: yes should I remove ?

File: mapping.ts (hunk 6 above)
Comment (viraj124) [reply mapping.ts:38]: the new one was in the branch you had earlier I think

File: yarn.lock (hunk 7 above)
Comment (viraj124) [reply yarn.lock:229]: this is just the internal open zepplin deps that come with the installation of that package think it should be fine

File: contracts/RentNft.sol (hunk 1 above)
Comment (nazariyv) [reply contracts/RentNft.sol:142]: fixed, will commit soon

File: contracts/mocks/PaymentToken.sol (hunk 4 above)
Comment (nazariyv) [reply contracts/mocks/PaymentToken.sol:8]: added

File: mapping.ts (hunk 6 above)
Comment (nazariyv) [reply mapping.ts:43]: updates now

**Review by viraj124 — COMMENTED** [review]
Looks Good

**Review by nazariyv — DISMISSED** [review]

**Comment by nazariyv** [comment]
! Please rebase this change when adding to main (not squash and merge, rebase). Thank you

File: contracts/mocks/TestERC1155.sol (hunk 5 above)
Comment (nazariyv) [reply contracts/mocks/TestERC1155.sol]: I removed

File: contracts/RentNftAddressProvider.sol (hunk 3 above)
Comment (CodinMaster) [reply contracts/RentNftAddressProvider.sol:43]: yup it's fine, already supports major stablecoins.

File: contracts/RentNft.sol [hunk 8]
```diff
@@ -1,32 +1,35 @@
...
 
+  // TODO: if there are defaults, mark the address to forbid from renting
   event Lent(
+    uint256 lentIndex,
     address indexed nftAddress,
     uint256 indexed tokenId,
     address indexed lender,
     uint256 maxDuration,
-    uint256 borrowPrice,
+    uint256 dailyPrice,
```
Comment (nazariyv) [inline contracts/RentNft.sol:24]: dailyBorrowPrice? @CodinMaster

File: contracts/RentNft.sol [hunk 9]
```diff
@@ -36,70 +39,73 @@ contract RentNft is ReentrancyGuard, Ownable, ERC721Holder {
...
-    uint256 actualDuration; // set by borrower. actual duration borrower will have the NFT for. This gets populated on the rent call
-    uint256 borrowPrice; // set by lender. how much the borrower has to pay irrevocably daily
-    uint256 borrowedAt; // set by borrower. time at which nft is borrowed
-    uint256 nftPrice; // set by lender. how much lender will receive as collateral if borrower does not return nft in time
+    address nftAddress;
+    uint256 tokenId;
+    uint256 maxDuration; // max borrow duration in days
+    uint256 dailyPrice; // how much the borrower has to pay irrevocably daily (per nft)
+    uint256 nftPrice; // how much lender will receive as collateral if borrower does not return nft in time
+    bool isRented;
```
Comment (nazariyv) [inline contracts/RentNft.sol:57]: we should stick to either "borrowing" or "renting" to avoid confusion

File: mapping.ts [hunk 10]
```diff
@@ -7,14 +7,15 @@ import {
...
+  Listing,
+  Rental,
   User,
   Approval as ApprovalSchema,
   ApprovedAll
 } from "./generated/schema";
 
-const createUser = (id: string): User => {
-  const user = new User(id);
+let createUser = (id: string): User => {
```
Comment (nazariyv) [inline mapping.ts:17]: why `let`? This function will not be re-assigned

File: mapping.ts [hunk 11]
```diff
@@ -7,14 +7,15 @@ import {
...
+  let user = new User(id);
```
Comment (nazariyv) [inline mapping.ts:18]: same here, the user is not being reassigned, why not use `const`?

File: mapping.ts [hunk 12]
```diff
@@ -23,37 +24,37 @@ const createUser = (id: string): User => {
...
+//   return nft;
+// };
 
 // ! notes for self
 // 1. string templating does not work
 // 2. variables from function scope not visible inside of .filter
 // 3. pushing directly into arrays won't work. Need to make a copy and then assign a copy to prop
 
-const getFaceId = (nftAddr: string, tokenId: string): string =>
+let getFaceId = (nftAddr: string, tokenId: string): string =>
```
Comment (nazariyv) [inline mapping.ts:39]: same here with lets, why let?

File: contracts/RentNft.sol (hunk 8 above)
Comment (viraj124) [reply contracts/RentNft.sol:24]: yeah @nazariyv since the owner enters per day price this name is more accurate

File: mapping.ts (hunk 10 above)
Comment (viraj124) [reply mapping.ts:17]: assembly script was giving an issue with const

File: contracts/RentNft.sol (hunk 8 above)
Comment (nazariyv) [reply contracts/RentNft.sol:24]: sorry, let me be clear. I am saying, let's rename it to `dailyBorrowPrice` then

Comment (nazariyv) [reply contracts/RentNft.sol:24]: also, let's stick to `Borrow` in place of `Rent` word everywhere, to stay consistent

File: contracts/RentNft.sol (hunk 9 above)
Comment (CodinMaster) [reply contracts/RentNft.sol:57]: ok, will update the bool `isRented`

File: contracts/RentNft.sol (hunk 8 above)
Comment (CodinMaster) [reply contracts/RentNft.sol:24]: gotcha will update

**Review by CodinMaster — APPROVED** [review]

**Review by viraj124 — COMMENTED** [review]
LGTM

**Comment by nazariyv** [comment]
total time spent on this reviewing and refactoring: 60-70 hours

**Pull request merged**